tiledb/bioimg/openslide.py

```diff
@@ -143,24 +139,64 @@ def get_best_level_for_downsample(self, factor: float) -> int:
         lla = np.where(np.array(self.level_downsamples) < factor)[0]
         return int(lla.max() if len(lla) > 0 else 0)
 
-    def _iter_level_dimensions(self) -> Iterator[Tuple[int, int]]:
-        for a in self._level_arrays:
-            dims = list(a.domain)
-            width = a.shape[dims.index(a.dim("X"))]
-            height = a.shape[dims.index(a.dim("Y"))]
-            yield width, height
-
     def _read_image(
         self,
         level: int,
-        dim_slice: Mapping[str, slice] = {},
+        dim_slice: MutableMapping[str, slice] = {},
         to_original_axes: bool = False,
         to_dask: bool = False,
     ) -> Union[np.ndarray, da.Array]:
-        tdb = self._level_arrays[level]
-        array = da.from_tiledb(tdb) if to_dask else tdb
-        dims = tuple(dim.name for dim in tdb.domain)
+        axes = Axes(self._group.meta["axes"] if to_original_axes else "YXC")
+        return self._levels[level].read(axes, dim_slice, to_dask)
+
+
+class TileDBOpenSlideLevel:
+    def __init__(self, uri: str):
+        self._tdb = tiledb.open(uri)
+
+    @property
+    def level(self) -> int:
+        return int(self._tdb.meta["level"])
+
+    @property
+    def dimensions(self) -> Tuple[int, int]:
+        a = self._tdb
+        dims = list(a.domain)
+        width = a.shape[dims.index(a.dim("X"))]
+        height = a.shape[dims.index(a.dim("Y"))]
+        return width // get_pixel_depth(a), height
+
+    @property
+    def properties(self) -> Mapping[str, Any]:
+        return dict(self._tdb.meta)
+
+    def read(
+        self,
+        axes: Axes,
+        dim_slice: MutableMapping[str, slice] = {},
+        to_dask: bool = False,
+    ) -> Union[np.ndarray, da.Array]:
+        dims = tuple(dim.name for dim in self._tdb.domain)
+        pixel_depth = get_pixel_depth(self._tdb)
+        if pixel_depth == 1:
+            axes_mapper = AxesMapper(Axes(dims), axes)
+        else:
+            x = dim_slice.get("X")
+            if x is not None:
+                dim_slice["X"] = slice(x.start * pixel_depth, x.stop * pixel_depth)
+            raise NotImplementedError
+
+        array = da.from_tiledb(self._tdb) if to_dask else self._tdb
         selector = tuple(dim_slice.get(dim, slice(None)) for dim in dims)
-        source_axes = Axes(dims)
-        target_axes = self.original_axes if to_original_axes else Axes("YXC")
-        return AxesMapper(source_axes, target_axes).map_array(array[selector])
+        return axes_mapper.map_array(array[selector])
+
+    def close(self) -> None:
+        self._tdb.close()
+
+
+def get_pixel_depth(obj: Union[tiledb.Array, tiledb.Filter]) -> int:
```

Comment:
Pixel depth is a property of the original image and should be extracted from the original file

Reply:
Indeed; putting this on draft until we get the channel order data from the original file.